### PR TITLE
[CDAP-14907] log file meta table migration

### PR DESCRIPTION
**Pull request by CuriousVini**
Migrating log file metadata store to new spi.

DUT: https://builds.cask.co/browse/CDAP-DUT6857
ITM: https://builds.cask.co/browse/IT-ITM-120

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 1]
```diff
@@ -848,4 +851,29 @@ public static void createTable(StructuredTableAdmin tableAdmin) throws IOExcepti
...
     }
   }
+
+  /**
+   * Schema for log file meta.
+   */
+  public static final class LogFileMetaStore {
+    public static final StructuredTableId LOG_FILE_META = new StructuredTableId("logfile_meta");
+
+    public static final String IDENTIFIER_FIELD = "identifier";
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: It would be good to add a comment as to what the identifier means.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java [hunk 2]
```diff
@@ -109,7 +107,6 @@ public void validate() throws InvalidPipelineException {
 
     Map<String, LogPipelineSpecification<T>> result = new HashMap<>();
     Set<String> checkpointPrefixes = new TreeSet<>();
-    checkpointPrefixes.add(RESERVED_CHECKPOINT_PREFIX);
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java:112]: Why is this removed?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java [hunk 3]
```diff
@@ -16,188 +16,114 @@
...
-        files.add(new LogLocation(LogLocation.VERSION_1,
-                                  Bytes.toLong(row.getRow(), prefixLength, Bytes.SIZEOF_LONG),
-                                  Bytes.toLong(row.getRow(), prefixLength + Bytes.SIZEOF_LONG, Bytes.SIZEOF_LONG),
-                                  // we store path in new format
-                                  Locations.getLocationFromAbsolutePath(locationFactory, (Bytes.toString(value))),
-                                  logPathIdentifier.getNamespaceId(), impersonator));
-
+    Range scanRange = Range.create(getKeyFields(logPathIdentifier.getRowkey(), 0L, 0L), Range.Bound.INCLUSIVE,
+                                   // end row-key is exclusive, so use endTimestamp + 1
+                                   getKeyFields(logPathIdentifier.getRowkey(), endTimestampMs + 1, 0L),
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java]: If you use `Range.Bound.EXCLUSIVE`, then there is no need to add 1.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java [hunk 4]
```diff
@@ -16,188 +16,114 @@
...
   }
 
-  private byte[] getOldRowKey(LogPathIdentifier logPathIdentifier) {
-    return Bytes.add(LoggingStoreTableUtil.OLD_FILE_META_ROW_KEY_PREFIX, logPathIdentifier.getOldRowkey().getBytes());
+  @SuppressWarnings("ConstantConditions")
+  private LogLocation fromRow(StructuredRow row, String namespace) {
+    return new LogLocation(LogLocation.VERSION_1,
+                           row.getLong(StoreDefinition.LogFileMetaStore.EVENT_TIME_FIELD),
+                           row.getLong(StoreDefinition.LogFileMetaStore.CREATION_TIME_FIELD),
+                           // we store path in new format
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java]: This comment is no longer required

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java [hunk 5]
```diff
@@ -97,61 +54,59 @@ public void run(DatasetContext context) throws Exception {
    * @param transactionTimeout transaction timeout to use for scanning entries, deleting entries.
    * @return list of DeleteEntry - used to get files to delete for which metadata has already been deleted
    */
+  @SuppressWarnings("ConstantConditions")
   public List<DeletedEntry> scanAndGetFilesToDelete(final long tillTime, final int transactionTimeout) {
     final List<DeletedEntry> toDelete = new ArrayList<>();
 
     // we make sure transactionTimeout is greater than TX_TIMEOUT_DISCOUNT_SECS in CDAPLogAppender check.
-    final int cutOffTransactionTime = transactionTimeout - TX_TIMEOUT_DISCOUNT_SECS;
+    final int cutOffTransactionTime = transactionTimeout - Constants.Logging.TX_TIMEOUT_DISCOUNT_SECS;
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java]: It would be better to use a batch cutoff based on a number, rather than time now since SPI may have different transaction timeouts based on the implementation being used.

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/file/TestFileLogging.java [hunk 6]
```diff
@@ -110,6 +113,10 @@ protected void configure() {
     txManager = injector.getInstance(TransactionManager.class);
     txManager.startAndWait();
 
+    StructuredTableRegistry structuredTableRegistry = injector.getInstance(StructuredTableRegistry.class);
+    structuredTableRegistry.initialize();
+    StoreDefinition.createAllTables(injector.getInstance(StructuredTableAdmin.class), structuredTableRegistry);
```
Comment (poornachandra) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/file/TestFileLogging.java]: Would be good to only create tables needed for this test

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/system/CDAPLogAppenderTest.java [hunk 7]
```diff
@@ -112,6 +114,10 @@ protected void configure() {
 
     txManager = injector.getInstance(TransactionManager.class);
     txManager.startAndWait();
+
+    StructuredTableRegistry structuredTableRegistry = injector.getInstance(StructuredTableRegistry.class);
+    structuredTableRegistry.initialize();
+    StoreDefinition.createAllTables(injector.getInstance(StructuredTableAdmin.class), structuredTableRegistry);
```
Comment (poornachandra) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/appender/system/CDAPLogAppenderTest.java]: Same here, and in other places below

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/clean/FileMetadataCleanerTest.java [hunk 8]
```diff
@@ -440,7 +298,52 @@ public void testFileMetadataWithCommonContextPrefix() throws Exception {
...
+  private boolean deleteAllMetaEntries(TransactionRunner transactionRunner) {
+    final List<DeletedEntry> deletedEntries = new ArrayList<>();
+    try {
+      TransactionRunners.run(transactionRunner, context -> {
+        Stopwatch stopwatch = new Stopwatch();
+        stopwatch.start();
+        StructuredTable table = context.getTable(StoreDefinition.LogFileMetaStore.LOG_FILE_META);
+        try (CloseableIterator<StructuredRow> iter = table.scan(Range.all(), Integer.MAX_VALUE)) {
+          while (iter.hasNext()) {
+            if (stopwatch.elapsedTime(TimeUnit.SECONDS) > CUTOFF_TIME_TRANSACTION) {
```
Comment (poornachandra) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/clean/FileMetadataCleanerTest.java]: Do we need a cutoff time here? How many files does this test generate?

**Review by poornachandra — COMMENTED** [review]
@CuriousVini Just a few comments

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java (hunk 2 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java:112]: This is because it was used to differentiate between metadata and checkpoints as they were shared table. Now that we have separated them out. It is not reserved checkpoint prefix any more.

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/clean/FileMetadataCleanerTest.java (hunk 8 above)
Comment (CuriousVini) [reply cdap-watchdog/src/test/java/co/cask/cdap/logging/clean/FileMetadataCleanerTest.java]: This is only used for cleaning existing entries for next test so have not modified existing logic

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java]: I think we want to include endTimestampMs.

**Comment by CuriousVini** [comment]
Addressed comments and resolving merge conflicts on StoreDefinition and UpgradeTool

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 1 above)
Comment (poornachandra) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: Just call it logging context in that case

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java [hunk 9]
```diff
@@ -51,26 +48,24 @@ public FileMetadataCleaner(TransactionRunner transactionRunner) {
...
         StructuredTable table = context.getTable(StoreDefinition.LogFileMetaStore.LOG_FILE_META);
-        Stopwatch stopwatch = new Stopwatch().start();
 
         boolean reachedEnd = false;
         Range range = Range.all();
-        while (!reachedEnd) {
+        int count = 0;
+        // stop cleaning up if there are no files to delete or if we have reached batch size
+        while (!reachedEnd && count < fileCleanupBatchSize) {
           try (CloseableIterator<StructuredRow> iter = table.scan(range, Integer.MAX_VALUE)) {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java]: Can use fileCleanupBatchSize in the scan limit so that you don't have to count the entries here

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java [hunk 10]
```diff
@@ -119,22 +115,22 @@ public FileMetadataCleaner(TransactionRunner transactionRunner) {
...
-                                                      final int cutOffTransactionTimeout) {
+                                                      final int fileCleanupBatchSize) {
     final List<DeletedEntry> deletedEntries = new ArrayList<>();
     try {
       TransactionRunners.run(transactionRunner, context -> {
-        Stopwatch stopwatch = new Stopwatch().start();
         StructuredTable table = context.getTable(StoreDefinition.LogFileMetaStore.LOG_FILE_META);
 
+        int count = 0;
         for (DeletedEntry entry : toDeleteRows) {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java:115]: toDelete rows will always be less than fileCleanupBatchSize now, right?

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/clean/FileMetadataCleanerTest.java [hunk 11]
```diff
@@ -305,13 +300,13 @@ public void testFileMetadataWithCommonContextPrefix() throws Exception {
...
+      int batch = 100;
       TransactionRunners.run(transactionRunner, context -> {
-        Stopwatch stopwatch = new Stopwatch();
-        stopwatch.start();
+        int count = 0;
         StructuredTable table = context.getTable(StoreDefinition.LogFileMetaStore.LOG_FILE_META);
         try (CloseableIterator<StructuredRow> iter = table.scan(Range.all(), Integer.MAX_VALUE)) {
           while (iter.hasNext()) {
-            if (stopwatch.elapsedTime(TimeUnit.SECONDS) > CUTOFF_TIME_TRANSACTION) {
+            if (count == batch) {
```
Comment (poornachandra) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/clean/FileMetadataCleanerTest.java]: In the test we can delete all entries in one shot, right? I don't think the test will create millions of entries

**Review by poornachandra — COMMENTED** [review]
@CuriousVini Few comments to simplify the code.

Also where is the default value for `file.cleanup.batch.size` defined?

**Comment by CuriousVini** [comment]
resolved merge conflict on `StoreDefinition`. Have also addressed latest comments @poornachandra please take another pass when you get a chance, thanks!

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java [hunk 12]
```diff
@@ -61,11 +61,10 @@ public FileMetadataCleaner(TransactionRunner transactionRunner) {
 
         boolean reachedEnd = false;
         Range range = Range.all();
-        int count = 0;
         // stop cleaning up if there are no files to delete or if we have reached batch size
-        while (!reachedEnd && count < fileCleanupBatchSize) {
-          try (CloseableIterator<StructuredRow> iter = table.scan(range, Integer.MAX_VALUE)) {
-            while (count < fileCleanupBatchSize) {
+        while (!reachedEnd && toDelete.size() < fileCleanupBatchSize) {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java]: Let's move the starting of new transaction inside the while loop so that every scan is in a new transaction. This will prevent transaction timeout in the NoSQL case.

**Review by poornachandra — COMMENTED** [review]
@CuriousVini One last comment

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java [hunk 13]
```diff
@@ -56,21 +56,14 @@ public FileMetadataCleaner(TransactionRunner transactionRunner) {
...
-      TransactionRunners.run(transactionRunner, context -> {
-        StructuredTable table = context.getTable(StoreDefinition.LogFileMetaStore.LOG_FILE_META);
-
-        boolean reachedEnd = false;
-        Range range = Range.all();
-        // stop cleaning up if there are no files to delete or if we have reached batch size
-        while (!reachedEnd && toDelete.size() < fileCleanupBatchSize) {
-          try (CloseableIterator<StructuredRow> iter = table.scan(range, fileCleanupBatchSize)) {
+      boolean reachedEnd = false;
+      final Range[] range = {Range.all()};
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java]: Maybe use Optional here, it will make the code look cleaner

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java [hunk 14]
```diff
@@ -56,21 +56,14 @@ public FileMetadataCleaner(TransactionRunner transactionRunner) {
...
+      // stop cleaning up if there are no files to delete or if we have reached batch size
+      while (!reachedEnd && toDelete.size() < fileCleanupBatchSize) {
+        TransactionRunners.run(transactionRunner, context -> {
+          StructuredTable table = context.getTable(StoreDefinition.LogFileMetaStore.LOG_FILE_META);
+          try (CloseableIterator<StructuredRow> iter = table.scan(range[0], fileCleanupBatchSize)) {
             while (toDelete.size() < fileCleanupBatchSize) {
```
Comment (poornachandra) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataCleaner.java]: Need to add `iter.hasNext()` to the while loop condition.

**Review by poornachandra — COMMENTED** [review]
@CuriousVini Two minor comments

**Review by poornachandra — APPROVED** [review]

**Comment by CuriousVini** [comment]
Squashed commits and resolved conflict with `KafkaLogProcessorPipelineTest`

**Comment by CuriousVini** [comment]
Both builds are green. merging.
